File: Backend/LoyaltyCards.Application/Services/RegistrationValidationService.cs
using System.Text.Json;

public class RegistrationValidationRules
{
    private static readonly Lazy<RegistrationValidationRules> _instance = new(() => LoadRules());
    public static RegistrationValidationRules Instance => _instance.Value;

    public PasswordRules Password { get; set; } = null!;
    public EmailRules Email { get; set; } = null!;

    private static RegistrationValidationRules LoadRules()
    {
        try
        {
            var basePath = AppDomain.CurrentDomain.BaseDirectory;
            var projectRoot = Path.GetFullPath(Path.Combine(basePath, "../../../../../"));
            var filePath = Path.Combine(projectRoot, "Shared", "validation-rules.json");

            Console.WriteLine($"Looking for validation rules at: {filePath}");

            if (!File.Exists(filePath))
                throw new FileNotFoundException($"Validation rules file not found at: {filePath}");

            var json = File.ReadAllText(filePath);
            Console.WriteLine($"Loaded JSON: {json}");

            var options = new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true,
                WriteIndented = true
            };

            var rules = JsonSerializer.Deserialize<RegistrationValidationRules>(json, options);

            if (rules?.Password == null || rules?.Email == null)
                throw new InvalidOperationException("Validation rules were not properly deserialized");

            return rules;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error loading validation rules: {ex}");
            throw;
        }
    }

    public class PasswordRules
    {
        public int MinLength { get; set; }
        public bool RequireUppercase { get; set; }
        public bool RequireLowercase { get; set; }
        public bool RequireDigits { get; set; }
        public bool RequireSpecialCharacters { get; set; }
        public string SpecialCharacters { get; set; } = string.Empty;
        public Dictionary<string, string> ErrorMessages { get; set; } = new();
    }

    public class EmailRules
    {
        public string Pattern { get; set; } = string.Empty;
        public Dictionary<string, string> ErrorMessages { get; set; } = new();
    }
}

File: Mobile/mobile/lib/services/registration_validation_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../api_config.dart';

class RegistrationValidationRules {
  static RegistrationValidationRules? _instance;
  static Future<RegistrationValidationRules> get instance async {
    _instance ??= await _loadRules();
    return _instance!;
  }

  final PasswordRules password;
  final EmailRules email;

  RegistrationValidationRules({required this.password, required this.email});

  static Future<RegistrationValidationRules> _loadRules() async {
    final response = await http.get(ApiConfig.validationRulesEndpoint());
    
    if (response.statusCode != 200) {
      throw Exception('Failed to load validation rules');
    }

    final json = jsonDecode(response.body);
    return RegistrationValidationRules(
      password: PasswordRules.fromJson(json['password']),
      email: EmailRules.fromJson(json['email']),
    );
  }
}

class PasswordRules {
  final int minLength;
  final bool requireUppercase;
  final bool requireLowercase;
  final bool requireDigits;
  final bool requireSpecialCharacters;
  final String specialCharacters;
  final Map<String, String> errorMessages;

  PasswordRules.fromJson(Map<String, dynamic> json)
      : minLength = json['minLength'],
        requireUppercase = json['requireUppercase'],
        requireLowercase = json['requireLowercase'],
        requireDigits = json['requireDigits'],
        requireSpecialCharacters = json['requireSpecialCharacters'],
        specialCharacters = json['specialCharacters'],
        errorMessages = Map<String, String>.from(json['errorMessages']);
}

class EmailRules {
  final String pattern;
  final Map<String, String> errorMessages;

  EmailRules.fromJson(Map<String, dynamic> json)
      : pattern = json['pattern'],
        errorMessages = Map<String, String>.from(json['errorMessages']);
}